File: go/payouts.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type PayoutRequest struct {
	Name          string `json:"name" binding:"required"`
	AccountNumber string `json:"accountNumber" binding:"required"`
}

type PayoutResponse struct {
	Message   string  `json:"message"`
	Reference *string `json:"reference,omitempty"`
}

type QuoteRequest struct {
	Source           string `json:"source"`
	FromAsset        string `json:"fromAsset"`
	ToCurrency       string `json:"toCurrency"`
	SettlementAmount int    `json:"settlementAmount"`
}

type InitializeRequest struct {
	QuoteID       string      `json:"quoteId"`
	CustomerID    string      `json:"customerId"`
	Country       string      `json:"country"`
	Reference     string      `json:"reference"`
	PaymentReason string      `json:"paymentReason"`
	Beneficiary   Beneficiary `json:"beneficiary"`
}

type Beneficiary struct {
	Type          string `json:"type"`
	AccountName   string `json:"accountName"`
	BankName      string `json:"bankName"`
	AccountNumber string `json:"accountNumber"`
}

type FinalizeRequest struct {
	QuoteID string `json:"quoteId"`
}

type ApiResponse struct {
	Data QuoteData `json:"data"`
}

type QuoteData struct {
	QuoteID string `json:"quoteId"`
}

func createPayout(c *gin.Context) {
	var payload PayoutRequest
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, PayoutResponse{
			Message: fmt.Sprintf("Invalid request: %s", err.Error()),
		})
		return
	}

	payoutReference := uuid.New().String()

	// Get environment variables
	apiURL := os.Getenv("API_URL")
	secretKey := os.Getenv("BITNOB_SECRET_KEY")

	if apiURL == "" || secretKey == "" {
		c.JSON(http.StatusInternalServerError, PayoutResponse{
			Message:   "Missing API configuration",
			Reference: &payoutReference,
		})
		return
	}

	// Step 1: Get quote
	quoteRequest := QuoteRequest{
		Source:           "offchain",
		FromAsset:        "usdt",
		ToCurrency:       "ngn",
		SettlementAmount: 100000,
	}

	quoteData, err := makeAPICall(fmt.Sprintf("%s/payouts/quotes", apiURL), secretKey, quoteRequest)
	if err != nil {
		fmt.Printf("Quote request failed: %v\n", err)
		c.JSON(http.StatusInternalServerError, PayoutResponse{
			Message:   fmt.Sprintf("Quote failed: %s", err.Error()),
			Reference: &payoutReference,
		})
		return
	}

	var quote ApiResponse
	if err := json.Unmarshal(quoteData, &quote); err != nil {
		fmt.Printf("Failed to parse quote response: %v\n", err)
		c.JSON(http.StatusInternalServerError, PayoutResponse{
			Message:   "Failed to parse quote response",
			Reference: &payoutReference,
		})
		return
	}

	fmt.Printf("[STEP 1] Quote Generated: %+v\n", quote)

	// Step 2: Initialize payout
	initializeRequest := InitializeRequest{
		QuoteID:       quote.Data.QuoteID,
		CustomerID:    "e22795d9-23f6-48e6-8b30-be5718abd876",
		Country:       "NG",
		Reference:     payoutReference,
		PaymentReason: "Bitnob Nigeria Faucet",
		Beneficiary: Beneficiary{
			Type:          "BANK",
			AccountName:   payload.Name,
			BankName:      "OPAY",
			AccountNumber: payload.AccountNumber,
		},
	}

	initData, err := makeAPICall(fmt.Sprintf("%s/payouts/initialize", apiURL), secretKey, initializeRequest)
	if err != nil {
		fmt.Printf("Initialize request failed: %v\n", err)
		c.JSON(http.StatusInternalServerError, PayoutResponse{
			Message:   fmt.Sprintf("Initialize failed: %s", err.Error()),
			Reference: &payoutReference,
		})
		return
	}

	fmt.Printf("[STEP 2] Quote Initialized: %s\n", string(initData))

	// Step 3: Finalize payout
	finalizeRequest := FinalizeRequest{
		QuoteID: quote.Data.QuoteID,
	}

	finalizeData, err := makeAPICall(fmt.Sprintf("%s/payouts/finalize", apiURL), secretKey, finalizeRequest)
	if err != nil {
		fmt.Printf("Finalize request failed: %v\n", err)
		c.JSON(http.StatusInternalServerError, PayoutResponse{
			Message:   fmt.Sprintf("Finalize failed: %s", err.Error()),
			Reference: &payoutReference,
		})
		return
	}

	fmt.Printf("[STEP 3] Quote Finalized: %s\n", string(finalizeData))

	c.JSON(http.StatusOK, PayoutResponse{
		Message:   "Payout process started. Check the webhook site for final confirmation.",
		Reference: &payoutReference,
	})
}

func makeAPICall(url, secretKey string, payload interface{}) ([]byte, error) {
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", secretKey))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("API error (%d): %s", resp.StatusCode, string(body))
	}

	return body, nil
}

File: php/payouts.php
<?php

function generateUUID() {
    return sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
        mt_rand(0, 0xffff), mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0x0fff) | 0x4000,
        mt_rand(0, 0x3fff) | 0x8000,
        mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
    );
}

function createPayout() {
    header('Content-Type: application/json');
    
    // Parse JSON input
    $input = json_decode(file_get_contents('php://input'), true);
    
    if (!$input || !isset($input['name']) || !isset($input['accountNumber'])) {
        http_response_code(400);
        echo json_encode([
            'message' => 'Invalid request: name and accountNumber are required'
        ]);
        return;
    }
    
    $payoutReference = generateUUID();
    
    // Get environment variables
    $apiURL = getenv('API_URL');
    $secretKey = getenv('BITNOB_SECRET_KEY');
    
    if (!$apiURL || !$secretKey) {
        http_response_code(500);
        echo json_encode([
            'message' => 'Missing API configuration',
            'reference' => $payoutReference
        ]);
        return;
    }
    
    try {
        // Step 1: Get quote
        $quoteRequest = [
            'source' => 'offchain',
            'fromAsset' => 'usdt',
            'toCurrency' => 'ngn',
            'settlementAmount' => 100000
        ];
        
        $quoteData = makeAPICall($apiURL . '/payouts/quotes', $secretKey, $quoteRequest);
        $quote = json_decode($quoteData, true);
        
        if (!$quote || !isset($quote['data']['quoteId'])) {
            throw new Exception('Invalid quote response');
        }
        
        error_log('[STEP 1] Quote Generated: ' . json_encode($quote));
        
        // Step 2: Initialize payout
        $initializeRequest = [
            'quoteId' => $quote['data']['quoteId'],
            'customerId' => 'e22795d9-23f6-48e6-8b30-be5718abd876',
            'country' => 'NG',
            'reference' => $payoutReference,
            'paymentReason' => 'Bitnob Nigeria Faucet',
            'beneficiary' => [
                'type' => 'BANK',
                'accountName' => $input['name'],
                'bankName' => 'OPAY',
                'accountNumber' => $input['accountNumber']
            ]
        ];
        
        $initData = makeAPICall($apiURL . '/payouts/initialize', $secretKey, $initializeRequest);
        error_log('[STEP 2] Quote Initialized: ' . $initData);
        
        // Step 3: Finalize payout
        $finalizeRequest = [
            'quoteId' => $quote['data']['quoteId']
        ];
        
        $finalizeData = makeAPICall($apiURL . '/payouts/finalize', $secretKey, $finalizeRequest);
        error_log('[STEP 3] Quote Finalized: ' . $finalizeData);
        
        http_response_code(200);
        echo json_encode([
            'message' => 'Payout process started. Check the webhook site for final confirmation.',
            'reference' => $payoutReference
        ]);
        
    } catch (Exception $e) {
        error_log('Payout failed: ' . $e->getMessage());
        http_response_code(500);
        echo json_encode([
            'message' => 'Payout failed: ' . $e->getMessage(),
            'reference' => $payoutReference
        ]);
    }
}

function makeAPICall($url, $secretKey, $payload) {
    $ch = curl_init();
    
    $headers = [
        'Authorization: Bearer ' . $secretKey,
        'Content-Type: application/json'
    ];
    
    curl_setopt_array($ch, [
        CURLOPT_URL => $url,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode($payload),
        CURLOPT_HTTPHEADER => $headers,
        CURLOPT_TIMEOUT => 30,
        CURLOPT_SSL_VERIFYPEER => true,
        CURLOPT_SSL_VERIFYHOST => 2
    ]);
    
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    
    curl_close($ch);
    
    if ($error) {
        throw new Exception('cURL error: ' . $error);
    }
    
    if ($httpCode >= 400) {
        throw new Exception('API error (' . $httpCode . '): ' . $response);
    }
    
    return $response;
}
?>